### DISCUSSION TITLE
Issue with Function Input in AgentGroupChat with Multiple Agents Using Python

### DISCUSSION THREAD

**Discussion by lovedeepatsgit** [microsoft/semantic-kernel#9955] [Q&A] [answered]
Hi everyone,

I'm encountering an issue while working with AgentGroupChat in Semantic Kernel. When I use only one chat (single agent), the function calling works perfectly, and the input provided to the function is accurate and as expected.

However, when I increase the number of agents to two or more, the input values passed to the function become incorrect and arbitrary, leading to unexpected results. It seems like the coordination between agents may be introducing inconsistencies in how inputs are generated or passed to the functions.

Has anyone else faced this issue, or does anyone have suggestions on how to ensure accurate function input when using multiple agents in AgentGroupChat? Any help would be greatly appreciated!

Thank you!

This is a sample structure of the code Below
----
from azure.core.credentials import AzureKeyCredential
from azure.search.documents import SearchClient
import os
from dotenv import load_dotenv
from typing import List
from semantic_kernel.kernel import Kernel
from semantic_kernel.functions import kernel_function
from semantic_kernel.connectors.ai.open_ai import AzureChatCompletion
from semantic_kernel.contents.chat_history import ChatHistory
from semantic_kernel.agents import ChatCompletionAgent, AgentGroupChat
from semantic_kernel.functions.kernel_arguments import KernelArguments
from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
from semantic_kernel.contents.chat_message_content import ChatMessageContent
from semantic_kernel.contents.utils.author_role import AuthorRole
from semantic_kernel.functions.kernel_function_from_prompt import KernelFunctionFromPrompt
from semantic_kernel.agents.strategies.selection.kernel_function_selection_strategy import KernelFunctionSelectionStrategy
from semantic_kernel.agents.strategies.termination.kernel_function_termination_strategy import KernelFunctionTerminationStrategy
import asyncio
import json
from typing import Annotated

# Placeholder for loading sensitive configuration
def load_config(file_path="config.json"):
    with open(file_path, "r") as config_file:
        return json.load(config_file)

# Load sensitive configuration
config = load_config()

azure_ai_endpoint = "<YOUR_AZURE_ENDPOINT>"
azure_ai_api_key = "<YOUR_API_KEY>"
azure_ai_model = "gpt-4o"
ai_search_endpoint = "<YOUR_SEARCH_ENDPOINT>"
ai_search_key = "<YOUR_SEARCH_KEY>"

load_dotenv()

class Retrieval:
    def __init__(self):
        service_endpoint = ai_search_endpoint
        api_key = ai_search_key
        index_name = "hmh-blob-w-schedule"

        self.search_client = SearchClient(
            endpoint=service_endpoint,
            index_name=index_name,
            credential=AzureKeyCredential(api_key),
        )

    @kernel_function(description="""
        This function retrieves data related to a specific educational standard or multiple standards.

        ### Input Structure:
        - Input format: "Generate Lesson Plan for Standard (standard) for grade [grade]".
        - Extract the value inside the parenthesis `(standard)` and use it as the input for this function.
        - Use the exact value without additional formatting.
    """)
    def hmh(self, query: Annotated[str, "Provide the exact standard code(s) extracted from the parenthesis in the query."]) -> List[str]:
        search_options = {
            "query_type": "semantic",
            "semantic_configuration_name": "azureml-default",
            "top": 5,
        }

        search_results = self.search_client.search(query, **search_options)
        results = [result["content"] for result in search_results]

        return tuple(results)

planner_kernel = Kernel()
planner_chat_completion_service = AzureChatCompletion(
    service_id="planner_agent_service",
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint,
)
planner_kernel.add_service(planner_chat_completion_service)

planner_agent = ChatCompletionAgent(
    service_id="planner_agent_service",
    kernel=planner_kernel,
    name="planner_agent",
    description="Generates a 4-day lesson plan using Retrieval Plugin.",
)

teacher_kernel = Kernel()
teacher_chat_completion_service = AzureChatCompletion(
    service_id="teacher_1_service",
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint,
)
teacher_kernel.add_service(teacher_chat_completion_service)

teacher_agent = ChatCompletionAgent(
    service_id="teacher_1_service",
    kernel=teacher_kernel,
    name="teacher_1_agent",
    description="Creates detailed content for Day 1 of the lesson plan.",
)

multi_chat = AgentGroupChat(
    agents=[planner_agent, teacher_agent],
    termination_strategy=KernelFunctionTerminationStrategy(
        agents=[planner_agent, teacher_agent],
        kernel=planner_kernel,
        history_variable_name="history",
        maximum_iterations=6,
    ),
)

async def async_output(standard: str, grade: str, kagan: str) -> str:
    messages = f"Generate Lesson Plan for Standard ({standard}) for grade {grade}"
    if kagan.lower() != "yes" and kagan.lower() != "no":
        return "Enter a valid value for Kagan: Yes or No"

    await multi_chat.add_chat_message(ChatMessageContent(role=AuthorRole.USER, content=messages))
    await multi_chat.reset()

    responses = []
    async for response in multi_chat.invoke():
        responses.append(response.content)
        responses.append("*" * 50)

    return "\n".join(responses)

# Example usage
result = asyncio.run(async_output("SC.5.E.5.1", "5", "No"))
print(result)

**Comment by moonbox3** [comment, marked as answer]
Hi @lovedeepatsgit, thank you for your question and for providing the details/context. A few questions for you:
- Can you provide more details around `However, when I increase the number of agents to two or more, the input values passed to the function become incorrect and arbitrary, leading to unexpected results.`. Do you have a chat history that shows this "weirdness?"
- In your sample code, I don't see any reference to adding your plugin. I want to confirm you are setting up the agent that uses the plugin in the following fashion: https://github.com/microsoft/semantic-kernel/blob/e66883e72a1c05a545705bf3427d0b4077ec489d/python/samples/getting_started_with_agents/step2_plugins.py#L74C5-L85C6

I'd love to see the chat history which can help provide some insight into how the chat history is managed between the two agents. If a tool call is required by one agent, that one agent should run the auto tool logic until it has a result, and then come back so that the other agent has a chance to run (depending upon the selection strategy or termination strategy).

**Reply by lovedeepatsgit** [reply]
Hi, @moonbox3 , first of all thanks for replying.

1. "However, when I increase the number of agents to two or more, the input values passed to the function become incorrect and arbitrary, leading to unexpected results." .. From this line I mean that for example if Typed the query "Generate a Lesson Plan for Standard MA.1.NSO.1.1 for grade 1", then the assigned agent should use "MA.1.NSO.1.1" in the defined plugin's attribute, and it is doing the exact when i use it single, but when I change the same to two agents at the place of MA.1.NSO.1.1 its starts providing something arbitrary like MAELLAA1, something like this. Below are some images to show this.
![Error_for_ELA 1 R 1 1_Same_Again_4](https://github.com/user-attachments/assets/5f50c6be-4829-451b-9467-a6128b5a34a4)
2. And here I am providing you the whole code I am using.
---
from azure.core.credentials import AzureKeyCredential
from azure.search.documents import SearchClient
import os
from dotenv import load_dotenv
from typing import List
from semantic_kernel.kernel import Kernel
from semantic_kernel.functions import kernel_function
from semantic_kernel.connectors.ai.open_ai import AzureChatCompletion
from semantic_kernel.contents.chat_history import ChatHistory
from semantic_kernel.agents import ChatCompletionAgent
from semantic_kernel.connectors.ai.open_ai.prompt_execution_settings.azure_chat_prompt_execution_settings import AzureChatPromptExecutionSettings
from semantic_kernel.connectors.ai.function_choice_behavior import FunctionChoiceBehavior
from semantic_kernel.functions.kernel_arguments import KernelArguments
from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
from semantic_kernel.contents.chat_message_content import ChatMessageContent
from semantic_kernel.contents.utils.author_role import AuthorRole
from semantic_kernel.functions.kernel_function_from_prompt import KernelFunctionFromPrompt
from semantic_kernel.agents.strategies.selection.kernel_function_selection_strategy import KernelFunctionSelectionStrategy
from semantic_kernel.agents.strategies.termination.kernel_function_termination_strategy import KernelFunctionTerminationStrategy
from semantic_kernel.agents import ChatCompletionAgent,AgentGroupChat
import asyncio
from promptflow import tool
import json
from pprint import pprint
from typing import Annotated

def load_config(file_path='Sensitivedata.json'):
    with open(file_path, 'r') as config_file:
        return json.load(config_file)

# Load the sensitive data from config.json
config = load_config()

# Access the data as dictionary values
azure_ai_endpoint = config.get("ai_endpoint")
azure_ai_api_key = config.get("ai_api_key")
azure_ai_model = 'gpt-4o'
ai_search_endpoint = config.get("ai_search_endpoint")
ai_search_key = config.get("ai_search_key")


load_dotenv()
class Retrieval:
    def __init__(self):
        service_endpoint = ai_search_endpoint
        api_key = ai_search_key
        index_name = "hmh-blob-w-schedule"

        self.search_client = SearchClient(
            endpoint=service_endpoint, 
            index_name=index_name, 
            credential=AzureKeyCredential(api_key)
        )

    @kernel_function(description="""
                     This function retrieves data related to a specific educational standard or multiple standards.

                    ### Input Structure:
                    - The input query will be provided in the format:
                    "Generate Lesson Plan for Standard (standard) for grade [grade]".

                    ### Guidelines:
                    - Extract the value inside the parenthesis `(standard)` and use it as the input for this function.
                    - Ensure you **only use the exact value from the parenthesis** without any additional context or formatting.

                    ### Common Mistakes to Avoid:
                    - Do not process or alter the extracted standard value.
                     """)
    def hmh(self, query: Annotated[str,"Provide the exact standard code(s) extracted from the parenthesis in the query."]) -> List[str]:
        query = {}
        search_options = {
            "query_type": "semantic", 
            "semantic_configuration_name": "azureml-default",
            "top": 5
        }

        search_results = self.search_client.search(query, **search_options)

        results = []
        for result in search_results:
            results.append(result['content'])

        return tuple(results)
    
class Retrieval_Kagan:
    def __init__(self):
        service_endpoint = ai_search_endpoint
        api_key = ai_search_key
        index_name = "kagan-image-index"

        self.search_client = SearchClient(
            endpoint=service_endpoint, 
            index_name=index_name, 
            credential=AzureKeyCredential(api_key)
        )

    @kernel_function(description="Performs a search on the study material to get content related to Kagan Activities. Always provide 'Kagan Coaching Form' as input")
    def kagan(self,query : str) -> List[str]:
        query="Kagan Coaching Form"
        search_options = {
            "top": 16
        }

        search_results = self.search_client.search(query, **search_options)

        results = []
        for result in search_results:
            results.append(result['merged_content'])

        return tuple(results)


planner_kernel = Kernel()
service_id = "planner_agent_service"
planner_chat_completion_service = AzureChatCompletion(
    service_id= service_id,
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint,
)
planner_kernel.add_service(planner_chat_completion_service)
settings = planner_kernel.get_prompt_execution_settings_from_service_id(service_id="planner_agent_service")
settings.function_choice_behavior = FunctionChoiceBehavior.Auto()

planner_agent = ChatCompletionAgent(
    service_id=service_id,
    kernel=planner_kernel,
    name="planner_agent",
    execution_settings=settings,
    description="Use to provide the initial plan for generation of Lesson Plan and by Taking Help from Retrieval Plugin.",
    instructions="""
                Your are an expert Azure Developer, your task is fetch the Data from Azure Ai Search for the asked Standard Only Send the asked Standard in the Argument of the Function for Better Results. And, After getting the result, You need to make an initial plan
                of 4-days, which will covers the short points used to guide the teachers for content generation. And for Each particular day you need to provide short 5 points which provides 
                the details of the topic to be covered on that day, the title of activity if the user asked for particularly Kagan Activity than use that other any activity can be used to
                teach on that day. And, try to increase the level of topics each day so that students can learn better without replication of topics. 
                """,
)
planner_kernel.add_plugin(plugin=Retrieval(), plugin_name="Retrieval_Plugin")

kagan_kernel = Kernel()
service_id = "kagan_agent_service"
kagan_chat_completion_service = AzureChatCompletion(
    service_id= service_id,
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint,
)
kagan_kernel.add_service(kagan_chat_completion_service)
settings = kagan_kernel.get_prompt_execution_settings_from_service_id(service_id="kagan_agent_service")
settings.function_choice_behavior = FunctionChoiceBehavior.Auto()

kagan_agent = ChatCompletionAgent(
    service_id=service_id,
    kernel=kagan_kernel,
    name="kagan_agent",
    execution_settings=settings,
    description="To fetch Kagan Activities for Teaching from Azure Ai Search.",
    instructions="""
                You task is to data from Azure Ai Search regarding Kagan Coaching Form. After Getting the Data You need to select 4 activities from all the activities randomly. And, after
                    selecting these you need to make an initial Activity Plan for Four Days which Should in the below format:
                    1. Should have the Heading 'Kagan Activities For Great Learning'
                    2. Activity for Day 1, with its Name and Description how the activity can be performed and that are the materials required for it in 5 to 8 Points.
                    3. Similarly as per above point provide details for Day 2, Day 3, and Day 4.
                """,
)
kagan_kernel.add_plugin(plugin=Retrieval_Kagan(), plugin_name="Kagan_Retrieval_Plugin")

teacher_1_kernel = Kernel()
service_id = "teacher_1_service"
teacher_1_chat_completion_service = AzureChatCompletion(
    service_id= service_id,
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint,
)
teacher_1_kernel.add_service(teacher_1_chat_completion_service)
settings = teacher_1_kernel.get_prompt_execution_settings_from_service_id(service_id="teacher_1_service")
settings.function_choice_behavior = FunctionChoiceBehavior.Auto()

teacher_1_agent = ChatCompletionAgent(
    service_id=service_id,
    kernel=teacher_1_kernel,
    name="teacher_1_agent",
    execution_settings=settings,
    description="Teacher Assigned for Generation of Lesson Plan for Day 1",
    instructions=""" 
                You are an expert educator tasked with developing a detailed lesson plan for a specific grade and standard. You have to create Lesson Plan for the first day of a 4-day curriculum. The students are weak in the specified standard and need additional time to study. The 4-day curriculum has already been planned, and now you need to prepare detailed content for Day 1, which should include: 

            1. **Activity Task**:  
            
                -The title of the activity should reference the relevant Standard(s) requested by the user. 
                -Provide a complete explanation of the activity. 

            2. **Content**:  

            - Like a concept or specific teaching content (around 2000 words) 

            3. **Story**: 

            - Imaginary Story to build the interest of students (around 2500 words) 

            4. **Vocabulary Builder**:  

            - Meanings of 5 words related to today's lesson 

            5. **Homework Sample**: 

            - Homework, along with its answers so that teacher get a reference to check the homework of students. 
            - First provide the questions only, after that provide the answer key.
            
            6. **Assessment**:  

            - A minimum of 10 questions (mix of subjective and objective) along their answer key .
            - Answer key should be after the questions.

            7.. **List of Websites**:  

            - Related to the concepts for this day, according to Kagan Coaching Form. 

            """,
)

teacher_2_kernel = Kernel()
service_id = "teacher_2_service"
teacher_2_chat_completion_service = AzureChatCompletion(
    service_id= service_id,
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint,
)
teacher_2_kernel.add_service(teacher_2_chat_completion_service)
settings = teacher_2_kernel.get_prompt_execution_settings_from_service_id(service_id="teacher_2_service")
settings.function_choice_behavior = FunctionChoiceBehavior.Auto()

teacher_2_agent = ChatCompletionAgent(
    service_id=service_id,
    kernel=teacher_2_kernel,
    name="teacher_2_agent",
    execution_settings=settings,
    description="Teacher Assigned for Generation of Lesson Plan for Day 2",
    instructions=""" 
                You are an expert educator tasked with developing a detailed lesson plan for a specific grade and standard. You have to create Lesson Plan for the second day of a 4-day curriculum. The students are weak in the specified standard and need additional time to study. The 4-day curriculum has already been planned, and now you need to prepare detailed content for Day 2, which should include: 

            1. **Activity Task**:  
            
                -The title of the activity should reference the relevant Standard(s) requested by the user. 
                -Provide a complete explanation of the activity. 

            2. **Content**:  

            - Like a concept or specific teaching content (around 2000 words) 

            3. **Story**: 

            - Imaginary Story to build the interest of students (around 2500 words) 

            4. **Vocabulary Builder**:  

            - Meanings of 5 words related to today's lesson 

            5. **Homework Sample**: 

            - Homework, along with its answers so that teacher get a reference to check the homework of students.
            - First provide the questions only, after that provide the answer key. 
            
            6. **Assessment**:  

            - A minimum of 10 questions (mix of subjective and objective) along their answer key.
            - Answer key should be after the questions.

            7.. **List of Websites**:  

            - Related to the concepts for this day, according to Kagan Coaching Form.
            
            """,
)

teacher_3_kernel = Kernel()
service_id = "teacher_3_service"
teacher_3_chat_completion_service = AzureChatCompletion(
    service_id= service_id,
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint,
)
teacher_3_kernel.add_service(teacher_3_chat_completion_service)
settings = teacher_3_kernel.get_prompt_execution_settings_from_service_id(service_id="teacher_3_service")
settings.function_choice_behavior = FunctionChoiceBehavior.Auto()

teacher_3_agent = ChatCompletionAgent(
    service_id=service_id,
    kernel=teacher_3_kernel,
    name="teacher_3_agent",
    execution_settings=settings,
    description="Teacher Assigned for Generation of Lesson Plan for Day 3",
    instructions=""" 
               You are an expert educator tasked with developing a detailed lesson plan for a specific grade and standard. You have to create Lesson Plan for the third day of a 4-day curriculum. The students are weak in the specified standard and need additional time to study. The 4-day curriculum has already been planned, and now you need to prepare detailed content for Day 3, which should include: 

            1. **Activity Task**:  
            
                -The title of the activity should reference the relevant Standard(s) requested by the user. 
                -Provide a complete explanation of the activity. 

            2. **Content**:  

            - Like a concept or specific teaching content (around 2000 words) 

            3. **Story**: 

            - Imaginary Story to build the interest of students (around 2500 words) 

            4. **Vocabulary Builder**:  

            - Meanings of 5 words related to today's lesson 

            5. **Homework Sample**: 

            - Homework, along with its answers so that teacher get a reference to check the homework of students.
            - First provide the questions only, after that provide the answer key. 
            
            6. **Assessment**:  

            - A minimum of 10 questions (mix of subjective and objective) along their answer key.
            - Answer key should be after the questions.

            7.. **List of Websites**:  

            - Related to the concepts for this day, according to Kagan Coaching Form. 
            
            """,
)

teacher_4_kernel = Kernel()
service_id = "teacher_4_service"
teacher_4_chat_completion_service = AzureChatCompletion(
    service_id= service_id,
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint,
)
teacher_4_kernel.add_service(teacher_4_chat_completion_service)
settings = teacher_4_kernel.get_prompt_execution_settings_from_service_id(service_id="teacher_4_service")
settings.function_choice_behavior = FunctionChoiceBehavior.Auto()

teacher_4_agent = ChatCompletionAgent(
    service_id=service_id,
    kernel=teacher_4_kernel,
    name="teacher_4_agent",
    execution_settings=settings,
    description="Teacher Assigned for Generation of Lesson Plan for Day 4",
    instructions=""" 
               You are an expert educator tasked with developing a detailed lesson plan for a specific grade and standard. You have to create Lesson Plan for the fourth day of a 4-day curriculum. The students are weak in the specified standard and need additional time to study. The 4-day curriculum has already been planned, and now you need to prepare detailed content for Day 4, which should include: 

            1. **Activity Task**:  
            
                -The title of the activity should reference the relevant Standard(s) requested by the user. 
                -Provide a complete explanation of the activity. 

            2. **Content**:  

            - Like a concept or specific teaching content (around 2000 words) 

            3. **Story**: 

            - Imaginary Story to build the interest of students (around 2500 words) 

            4. **Vocabulary Builder**:  

            - Meanings of 5 words related to today's lesson 

            5. **Homework Sample**: 

            - Homework, along with its answers so that teacher get a reference to check the homework of students.
            - First provide the questions only, after that provide the answer key. 
            
            6. **Assessment**:  

            - A minimum of 10 questions (mix of subjective and objective) along their answer key.
            - Answer key should be after the questions.

            7.. **List of Websites**:  

            - Related to the concepts for this day, according to Kagan Coaching Form. 
            
            """,
)

Assessment_kernel = Kernel()
service_id = "Assessment_service"
Assessment_chat_completion_service = AzureChatCompletion(
    service_id= service_id,
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint,
)
Assessment_kernel.add_service(Assessment_chat_completion_service)
settings = Assessment_kernel.get_prompt_execution_settings_from_service_id(service_id="Assessment_service")
settings.function_choice_behavior = FunctionChoiceBehavior.Auto()

Assessment_agent = ChatCompletionAgent(
    service_id=service_id,
    kernel=Assessment_kernel,
    name="Assessment_agent",
    execution_settings=settings,
    description="Teacher assigned for assessment generation at the end of the four-day lessson planning",
    instructions=""" 
            You are an expert educator with experience in creating assessments for students. Your task is to analyze the context and information 
            provided to you, which pertains to a specific grade and standard. Based on the teaching conducted over previous days, you need to create an 
            assessment comprising 50 questions. This assessment should include a mix of subjective and objective questions. Remember, this assessment will 
            be posted on an online portal, so avoid using phrases like "circle the correct answer." Instead, use "select the correct answer" to make the assessment more suitable for an online format.
            
            Reply 'TERMINATE' when the task is completed and final assessment has been created/
            """,
)

AGENT_ONE = "kagan_agent"
AGENT_TWO = "planner_agent"
AGENT_THREE = "teacher_1_agent"
AGENT_FOUR = "teacher_2_agent"
AGENT_FIVE = "teacher_3_agent"
AGENT_SIX = "teacher_4_agent"
AGENT_SEVEN = "Assessment_agent"

selection_function = KernelFunctionFromPrompt(
    function_name="selection",
    prompt=f"""
                Determine which participant takes the next turn in a conversation based on the the most recent participant.
                State only the name of the participant to take the next turn.
                No participant should take more than one turn in a row.

                Choose only from these participants:
                - {AGENT_ONE}
                - {AGENT_TWO}
                - {AGENT_THREE}
                - {AGENT_FOUR}
                - {AGENT_FIVE}
                - {AGENT_SIX}
                - {AGENT_SEVEN}

                Always follow these rules when selecting the next participant:
                - After user input, it is {AGENT_ONE}'s turn.
                - After {AGENT_ONE} replies, it is {AGENT_TWO}'s turn.
                - After {AGENT_TWO} replies, it is {AGENT_THREE}'s turn.
                - After {AGENT_THREE} replies, it is {AGENT_FOUR}'s turn.
                - After {AGENT_FOUR} replies, it is {AGENT_FIVE}'s turn.
                - After {AGENT_FIVE} replies, it is {AGENT_SIX}'s turn.
                - After {AGENT_SIX} replies, it is {AGENT_SEVEN}'s turn.
                - End the loop
                
                History:
                {{{{$history}}}}

            """
)

selection_function_2 = KernelFunctionFromPrompt(
    function_name="selection",
    prompt=f"""
                Determine which participant takes the next turn in a conversation based on the the most recent participant.
                State only the name of the participant to take the next turn.
                No participant should take more than one turn in a row.

                Choose only from these participants:
                - {AGENT_TWO}
                - {AGENT_THREE}
                - {AGENT_FOUR}
                - {AGENT_FIVE}
                - {AGENT_SIX}
                - {AGENT_SEVEN}

                Always follow these rules when selecting the next participant:
                - After user input, it is {AGENT_TWO}'s turn.
                - After {AGENT_TWO} replies, it is {AGENT_THREE}'s turn.
                - After {AGENT_THREE} replies, it is {AGENT_FOUR}'s turn.
                - After {AGENT_FOUR} replies, it is {AGENT_FIVE}'s turn.
                - After {AGENT_FIVE} replies, it is {AGENT_SIX}'s turn.
                - After {AGENT_SIX} replies, it is {AGENT_SEVEN}'s turn.
                - End the loop
                History:
                {{{{$history}}}}

            """
)

TERMINATION_KEYWORD = "TERMINATE"

termination_function = KernelFunctionFromPrompt(
    function_name="termination_function",
    prompt="""
            If last reply is by {AGENT_SEVEN}. End the loop.
            
            RESPONSE:
            {{{{$history}}}}

            """
)

termination_kernel = Kernel()
termination_chat_completion_service = AzureChatCompletion(
    service_id="termination",
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint
)
termination_kernel.add_service(termination_chat_completion_service)

selection_kernel = Kernel()
selection_chat_completion_service = AzureChatCompletion(
    service_id="selection",
    api_key=azure_ai_api_key,
    deployment_name=azure_ai_model,
    endpoint=azure_ai_endpoint
)
selection_kernel.add_service(selection_chat_completion_service)

multi_chat = AgentGroupChat(
    agents=[kagan_agent,planner_agent,teacher_1_agent,teacher_2_agent,teacher_3_agent,teacher_4_agent,Assessment_agent],
    termination_strategy=KernelFunctionTerminationStrategy(
        agents=[kagan_agent,planner_agent,teacher_1_agent,teacher_2_agent,teacher_3_agent,teacher_4_agent,Assessment_agent],
        function=termination_function,
        kernel= termination_kernel,
        result_parser= lambda result: TERMINATION_KEYWORD in str(result.value[0]),
        history_variable_name="history",
        maximum_iterations=7,
    ),
    selection_strategy=KernelFunctionSelectionStrategy(
        function=selection_function,
        kernel=selection_kernel,
        result_parser= lambda result: str(result.value[0]) if result.value is not None else print('TERMINATE'),
        agent_variable_name="agents",
        history_variable_name="history",
    ),
)

multi_chat_2 = AgentGroupChat(
    agents=[planner_agent,teacher_1_agent,teacher_2_agent,teacher_3_agent,teacher_4_agent,Assessment_agent],
    termination_strategy=KernelFunctionTerminationStrategy(
        agents=[planner_agent,teacher_1_agent,teacher_2_agent,teacher_3_agent,teacher_4_agent,Assessment_agent],
        function=termination_function,
        kernel= termination_kernel,
        result_parser= lambda result: TERMINATION_KEYWORD in str(result.value[0]),
        history_variable_name="history",
        maximum_iterations=6,
    ),
    selection_strategy=KernelFunctionSelectionStrategy(
        function=selection_function_2,
        kernel=selection_kernel,
        result_parser= lambda result: str(result.value[0]) if result.value is not None else print('TERMINATE'),
        agent_variable_name="agents",
        history_variable_name="history",
    ),
)

@tool
async def async_output(standard: str, Grade: str, Kagan: str) -> str:

    messages =[f"Generate Lesson Plan for Standard ({standard}) for grade {Grade}"]
    delta = ["teacher_1_agent", "teacher_2_agent", "teacher_3_agent", "teacher_4_agent", "Assessment_agent"]

    if Kagan == 'Yes':
        question_2 = messages[0]
        responses = []
        await multi_chat.add_chat_message(ChatMessageContent(role=AuthorRole.USER, content=question_2))

        async for response in multi_chat.invoke():
            if response.name in delta:
                responses.append(response.content)
                responses.append("*" * 50)
        return "\n".join(responses)

    elif Kagan == "No":
        question_2 = messages[0]
        responses = []
        await multi_chat_2.add_chat_message(ChatMessageContent(role=AuthorRole.USER, content=question_2))

        async for response in multi_chat_2.invoke():
            if response.name in delta:
                responses.append(response.content)
                responses.append("*" * 50)
        return "\n".join(responses)

    else:
        return "Enter a valid value for Kagan: Yes or No"









----
I am trying to use semantic kernels inside prompt flow and its getting inputs directly from the input node.

**Reply by lovedeepatsgit** [reply]
Thanks @moonbox3  The issue has been resolved by adjusting the temperature of the planner agent using extention_data and providing the brief descriptions of the functions. Thanks for replying.

**Reply by moonbox3** [reply]
Thanks for letting us know, @lovedeepatsgit!
